### PR TITLE
Make font name comparison case-insensitive

### LINKED ISSUE #9360: Unable to load font X, using font X instead?!?
# Environment

```none
Windows build number: Win32NT             10.0.19042.0 Microsoft Windows NT 10.0.19042.0
Windows Terminal version (if applicable): 1.6.10571.0
```

# Steps to reproduce

Run Windows Terminal on my machine (sorry)

# Expected behavior

It runs normally

# Actual behavior

It displays this nonsense message

WindowsTerminal_BGD4W2SnLW

Was working until I let the Microsoft Store upgrade it 10 minutes ago

### PR DESCRIPTION
**Pull request by eugenesmlv**
Make font name comparison case-insensitive, so a warning message won't
be displayed when settings contain a font name with the wrong case.

Pros:
* The warning message isn't shown.

Cons:
* If one clicks the `Save` button on SUI, the fallback font will be
  applied. That happens because the selected font face in SUI is
  fallback font. 

Should we also silently change the font face in the settings in that case?

Fixes #9360

### CODE REVIEW

File: src/cascadia/TerminalControl/TermControl.cpp [hunk 1]
```diff
@@ -2111,13 +2111,20 @@ namespace winrt::Microsoft::Terminal::TerminalControl::implementation
         //      actually fail. We need a way to gracefully fallback.
         _renderer->TriggerFontChange(newDpi, _desiredFont, _actualFont);
 
+        const auto actualFontName = _actualFont.GetFaceName();
+        const auto desiredFontName = _desiredFont.GetFaceName();
+
+        auto caseInsensitiveEquality = [](wchar_t a, wchar_t b) { return ::towlower(a) == ::towlower(b); };
```
Comment (DHowett) [inline src/cascadia/TerminalControl/TermControl.cpp:2117]: Worried that this uses the user's locale for a not-locale-specific font name. Maybe we can use `CompareStringOrdinal` with `bIgnoreCase=TRUE` to get around that?

**Review by DHowett — APPROVED** [review]
Thanks!

**Comment by DHowett** [comment]
@msftbot make sure @miniksa signs off on this

**Comment by DHowett** [comment]
When you say that the fallback font is displayed in the settings UI, do you mean that it's not selected in the dropdown box because it can't be found in the list (case sensitively)? @carlos-zamora -- can you point @eugenesmlv to the right place?

**Review by miniksa — APPROVED** [review]

**Comment by carlos-zamora** [comment]
> When you say that the fallback font is displayed in the settings UI, do you mean that it's not selected in the dropdown box because it can't be found in the list (case sensitively)? @carlos-zamora -- can you point @eugenesmlv to the right place?

@eugenesmlv If fallback is happening for settings UI, you may want to do the same kind of comparison [here](https://github.com/microsoft/terminal/blob/99b09c08d51ba073889820ff8d520eeb353045f3/src/cascadia/TerminalSettingsEditor/Profiles.cpp#L431)

**Comment by DHowett** [comment]
Oh no, we're using the localized name there. I might have led you astray. Carlos, do you know if we use _localized_ or _raw_ font names when looking up fonts? This could .. be .. important ...

**Comment by eugenesmlv** [comment]
@DHowett yes, I do. When I open the settings editor the `profileFontFace` in `CurrentFontFace` method is set to the value from `settings.json`. I've tried to call `_settings.FontFace(actualFontName)` after the string comparison, but it doesn't help.

https://github.com/microsoft/terminal/blob/99b09c08d51ba073889820ff8d520eeb353045f3/src/cascadia/TerminalSettingsEditor/Profiles.cpp#L429-L442

**Comment by carlos-zamora** [comment]
> Oh no, we're using the localized name there. I might have led you astray. Carlos, do you know if we use _localized_ or _raw_ font names when looking up fonts? This could .. be .. important ...

Unfortunately, I actually don't know. I tracked it down to the renderer here:

https://github.com/microsoft/terminal/blob/99b09c08d51ba073889820ff8d520eeb353045f3/src/renderer/dx/DxFontRenderData.cpp#L643

And the docs for `IDWriteFontCollection::FindFamilyName` don't say anything on it, it seems:

https://docs.microsoft.com/en-us/windows/win32/api/dwrite/nf-dwrite-idwritefontcollection-findfamilyname

**Comment by miniksa** [comment]
I'll look behind the `FindFamilyName` interface and see where it's pulling from...

**Comment by DHowett** [comment]
@miniksa this is really gonna ruin your day.

image

**Comment by miniksa** [comment]
> @miniksa this is really gonna ruin your day.
> 
> image

NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN

**Comment by miniksa** [comment]
> I'll look behind the `FindFamilyName` interface and see where it's pulling from...

It actually contains all names in that dictionary but will yield the `en-us` one first if one exists... UGH.

**Comment by DHowett** [comment]
We might have a couple things to do here.

1. Control/Warning: We probably cannot take the "easy way out" with the font name comparison any more. We need the DX Engine to _tell TermControl that fallback took place_. That's the only way to know if we actually hit the provided fallback table.
2. Settings UI: We probably need to store every localized name (or maybe just the current localized name AND the base unlocalized name) in the SUI ViewModel Font object so that we can map the user's request into an actual font name.

**Comment by miniksa** [comment]
To get the list of names... we have to...

`IDWriteFontFamilyCollection::FindFamilyName` which will give us the `UINT32` of the index of the font family in that collection that matches the name. It will search its internal dictionary which contains a map of all the names it knows of to the indexes. And as an example both `MS Gothic` and `MS　ゴシック` should yield the same index.

Then we have to do `IDWriteFontFamilyCollection::GetFontFamily` with the index to get a `IDWriteFontFamily`.

Then `IDWriteFontFamily::GetFamilyNames` to get all the names that it knows about in all languages as `IDWriteLocalizedStrings`.

Then we need to loop `IDWriteLocalizedStrings::GetStringLength` and `IDWriteLocalizedStrings::GetString` for the count of `IDWriteLocalizedStrings::GetCount` to get all the possible names the family has in all locales, using `::GetLocaleName` and `::GetLocaleNameLength` if we want to know the matching locale name too.

**Comment by DHowett** [comment]
That'll be good for 2, the Settings UI work. It won't help us much for 1, the "we need to actually know about fallback instead of guessing based on the name" work.

**Comment by miniksa** [comment]
k let me dig to see if there's a better way to know of the fallback next.

**Comment by carlos-zamora** [comment]
@DHowett 

> 2. Settings UI: We probably need to store every localized name (or maybe just the current localized name AND the base unlocalized name) in the SUI ViewModel Font object so that we can map the user's request into an actual font name.

We already store the current localized name and the base unlocalized name in `Editor::Font`. However, we only compare what the user wrote to the localized name (see below).

https://github.com/microsoft/terminal/blob/99b09c08d51ba073889820ff8d520eeb353045f3/src/cascadia/TerminalSettingsEditor/Profiles.cpp#L429-L442

Should we just make that look something like this then:
```c++
 for (const auto& font : currentFontList) 
 { 
     // GH#XXXX: check localized name and base name
     if (font.LocalizedName() == profileFontFace || font.Name() == profileFontFace) 
     { 
         return box_value(font); 
     } 
     else if (font.LocalizedName() == L"Cascadia Mono" || font.Name() == L"Cascadia Mono") 
     { 
         fallbackFont = box_value(font); 
     } 
 } 
```

**Comment by miniksa** [comment]
> @DHowett
> 
> > 1. Settings UI: We probably need to store every localized name (or maybe just the current localized name AND the base unlocalized name) in the SUI ViewModel Font object so that we can map the user's request into an actual font name.
> 
> We already store the current localized name and the base unlocalized name in `Editor::Font`. However, we only compare what the user wrote to the localized name (see below).
> 
> https://github.com/microsoft/terminal/blob/99b09c08d51ba073889820ff8d520eeb353045f3/src/cascadia/TerminalSettingsEditor/Profiles.cpp#L429-L442
> 
> Should we just make that look something like this then:
> 
> ```c++
>  for (const auto& font : currentFontList) 
>  { 
>      // GH#XXXX: check localized name and base name
>      if (font.LocalizedName() == profileFontFace || font.Name() == profileFontFace) 
>      { 
>          return box_value(font); 
>      } 
>      else if (font.LocalizedName() == L"Cascadia Mono" || font.Name() == L"Cascadia Mono") 
>      { 
>          fallbackFont = box_value(font); 
>      } 
>  } 
> ```

There could be N names. Not just one localized and one unlocalized. Or at least that's how DWrite stores it internally. :|

**Comment by DHowett** [comment]
And we're going to want to make the comparison case-insensitively.

**Comment by DHowett** [comment]
@eugenesmlv sorry we took over your pull request as our chat channel for this issue. There's a significant Teams outage, it seems, and GitHub has really worked out for us 😉

**Comment by eugenesmlv** [comment]
@DHowett it was interesting! Unfortunately, I don't know much about the terminal internals to participate in the discussion, but I'll be glad to help you fix that font comparison issue ☺️

**Comment by miniksa** [comment]
> k let me dig to see if there's a better way to know of the fallback next.

We know that we fall back to a different face in `DxFontRenderData::_ResolveFontFaceWithFallback` after the first `_FindFontFace` call fails.

We also know that we couldn't find the name we were looking for in the current locale and did a fallback to a different name in `DxFontRenderData::_GetFontFamilyName` (and this function is also a great example of how to dig the names out like I was describing for the `IDWriteLocalizedStrings` business above.)

So I think just bubbling this information out of the lookups and to the interface is just what needs to be done.

**Comment by miniksa** [comment]
@eugenesmlv, thanks for this proposal. But I'm going to close this in favor of the solution I wrote in #9734 that handles a few more scenarios than this and avoids comparing strings at all. Thanks for lighting a fire under me. :)

**Pull request closed without merging**